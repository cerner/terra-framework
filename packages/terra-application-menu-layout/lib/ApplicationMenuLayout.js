"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _ApplicationMenuLayoutModule = _interopRequireDefault(require("./ApplicationMenuLayout.module.scss"));
var _excluded = ["content", "extensions", "footer", "header"];
var cx = _bind.default.bind(_ApplicationMenuLayoutModule.default);
var propTypes = {
  /**
   * Content element to be placed within the fill area of the header.
   * */
  content: _propTypes.default.element,
  /**
   * Extensions element to be placed before the end of the header.
   * */
  extensions: _propTypes.default.element,
  /**
   * Footer element to be placed at the end of the header.
   * */
  footer: _propTypes.default.element,
  /**
   * Header element to be placed at the start of the header.
   * */
  header: _propTypes.default.element
};
var ApplicationMenuLayout = function ApplicationMenuLayout(_ref) {
  var content = _ref.content,
    extensions = _ref.extensions,
    footer = _ref.footer,
    header = _ref.header,
    customProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var menuClassNames = cx(['menu', 'fill', customProps.className]);
  var headerElement;
  if (header) {
    headerElement = /*#__PURE__*/_react.default.createElement("div", {
      className: cx(['fit', 'header'])
    }, header);
  }
  var contentElement;
  if (content) {
    contentElement = /*#__PURE__*/_react.default.createElement("div", {
      className: cx(['fill', 'content'])
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: cx('normalizer')
    }, content));
  }
  var extensionsElement;
  if (extensions) {
    extensionsElement = /*#__PURE__*/_react.default.createElement("div", {
      className: cx(['fit', 'widget'])
    }, extensions);
  }
  var footerElement;
  if (footer) {
    footerElement = /*#__PURE__*/_react.default.createElement("div", {
      className: cx(['fit', 'footer'])
    }, footer);
  }
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, customProps, {
    className: menuClassNames
  }), headerElement, /*#__PURE__*/_react.default.createElement("div", {
    className: cx(['fill', 'body'])
  }, extensionsElement, contentElement), footerElement);
};
ApplicationMenuLayout.propTypes = propTypes;
var _default = ApplicationMenuLayout;
exports.default = _default;