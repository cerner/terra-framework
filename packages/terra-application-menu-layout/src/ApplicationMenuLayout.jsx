import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import AppDelegate from 'terra-app-delegate';

import styles from './ApplicationMenuLayout.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The AppDelegate instance provided by the containing component. If present, its properties will propagate to the children components.
   * */
  app: AppDelegate.propType,
  /**
   * Content element to be placed within the fill area of the header.
   * */
  content: PropTypes.element,
  /**
   * Extensions element to be placed beforethe end of the header.
   * */
  extensions: PropTypes.element,
  /**
   * Footer element to be placed at the end of the header.
   * */
  footer: PropTypes.element,
  /**
   * Header element to be placed at the start of the header.
   * */
  header: PropTypes.element,
  /**
   * Layout configuration to handle size and toggle functionality.
   * */
  layoutConfig: PropTypes.shape({
    size: PropTypes.string,
    toggleMenu: PropTypes.func,
    menuIsOpen: PropTypes.bool,
    togglePin: PropTypes.func,
    menuIsPinned: PropTypes.bool,
  }),
};

const defaultProps = {
  layoutConfig: {
    menuIsOpen: false,
    menuIsPinned: false,
  },
};

const appendPropsToElement = (app, size, element) => React.cloneElement(element, { app, size });

const ApplicationMenuLayout = ({
  app,
  content,
  extensions,
  footer,
  header,
  layoutConfig,
  ...customProps
}) => {
  const menuClassNames = cx([
    'menu',
    'fill',
    customProps.className,
  ]);

  let headerElement;
  if (header) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, header);
    headerElement = <div className={cx(['fit', 'header'])}>{clonedElement}</div>;
  }

  let contentElement;
  if (content) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, content);
    contentElement = <div className={cx(['fill', 'content'])}><div className={cx('normalizer')}>{clonedElement}</div></div>;
  }

  let extensionsElement;
  if (extensions) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, extensions);
    extensionsElement = <div className={cx(['fit', 'widget'])}>{clonedElement}</div>;
  }

  let footerElement;
  if (footer) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, footer);
    footerElement = <div className={cx(['fit', 'footer'])}>{clonedElement}</div>;
  }

  return (
    <div {...customProps} className={menuClassNames}>
      {headerElement}
      <div className={cx(['fill', 'body'])}>
        {extensionsElement}
        {contentElement}
      </div>
      {footerElement}
    </div>
  );
};

ApplicationMenuLayout.propTypes = propTypes;
ApplicationMenuLayout.defaultProps = defaultProps;

export default ApplicationMenuLayout;
