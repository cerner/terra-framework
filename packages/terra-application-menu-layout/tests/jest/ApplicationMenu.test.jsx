import React from 'react';

import ApplicationMenuLayout from '../../src/ApplicationMenuLayout';

describe('ApplicationMenuLayout', () => {
  it('should render without optional props', () => {
    const result = shallow((
      <ApplicationMenuLayout />
    ));
    expect(result).toMatchSnapshot();
  });

  it('should render with provided props', () => {
    const result = shallow((
      <ApplicationMenuLayout
        header={<div id="test-header">Header</div>}
        extensions={<div id="test-extensions">Extensions</div>}
        content={<div id="test-content">Content</div>}
        footer={<div id="test-footer">Footer</div>}
      />
    ));

    expect(result).toMatchSnapshot();
  });
});
