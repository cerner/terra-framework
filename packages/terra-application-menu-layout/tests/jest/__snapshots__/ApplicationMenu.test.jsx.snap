// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationMenuLayout should render with provided props 1`] = `
<div
  className="menu fill"
>
  <div
    className="fit header"
  >
    <div
      id="test-header"
    >
      Header
    </div>
  </div>
  <div
    className="fill body"
  >
    <div
      className="fit widget"
    >
      <div
        id="test-extensions"
      >
        Extensions
      </div>
    </div>
    <div
      className="fill content"
    >
      <div
        className="normalizer"
      >
        <div
          id="test-content"
        >
          Content
        </div>
      </div>
    </div>
  </div>
  <div
    className="fit footer"
  >
    <div
      id="test-footer"
    >
      Footer
    </div>
  </div>
</div>
`;

exports[`ApplicationMenuLayout should render without optional props 1`] = `
<div
  className="menu fill"
>
  <div
    className="fill body"
  />
</div>
`;
